homeassistant/components/melnor/models.py

```diff
@@ -0,0 +1,107 @@
+"""melnor integration models."""
+
+from datetime import timedelta
+import logging
+
+from melnor_bluetooth.device import Device
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import DEFAULT_CONNECTION_TIMEOUT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class MelnorDataUpdateCoordinator(DataUpdateCoordinator[Device]):
+    """Melnor data update coordinator."""
+
+    _device: Device
+    _has_active_connection: bool = False
+
+    def __init__(self, hass: HomeAssistant, device: Device) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Melnor Bluetooth",
+            update_interval=timedelta(seconds=5),
+        )
+        self._device = device
+
+    async def _async_update_data(self):
+        """Update data."""
+
+        if self._device.is_connected:
+
+            if not self._has_active_connection:
+                _LOGGER.debug("%s has re-connected", self._device.mac)
+
+            self._has_active_connection = True
+            await self._device.fetch_state()
+
+        # The melnor-bluetooth library handles exceptions for us.
+        # We just need to check the connection state and attempt to create a new connection if it drops.
+        else:
+
+            if self._has_active_connection:
+                self._has_active_connection = False
+                _LOGGER.warning("%s has disconnected", self._device.mac)
+
+                # Since we just became disconnected we'll return early to trigger an entity update ASAP to show the entities as unavailable.
+                # We update frequently and we'll start attempting to reconnect on the next pass in a few seconds.
+                return self._device
+
+            await self._device.connect(timeout=DEFAULT_CONNECTION_TIMEOUT)
+
+        return self._device
```

Comment:
It looks like quite a bit of the logic can be moved into the library instead.

Ideally hass doesn't know about the connection state beyond if the device is available or not and the library takes care of all of that.